### ISSUE TITLE
Build failure: ncdu on Darwin

### ISSUE THREAD

**Issue by al3xtjames** [NixOS/nixpkgs#287861]
### Steps To Reproduce

1. Run `nix build nixpkgs#ncdu` on Darwin

### Build log

```
@nix { "action": "setPhase", "phase": "unpackPhase" }
Running phase: unpackPhase
unpacking source archive /nix/store/s7smxibjail9wlambl6ylbi0ifvrpi2h-ncdu-2.3.tar.gz
source root is ncdu-2.3
setting SOURCE_DATE_EPOCH to timestamp 1691157970 of file ncdu-2.3/ncdu.1
@nix { "action": "setPhase", "phase": "patchPhase" }
Running phase: patchPhase
@nix { "action": "setPhase", "phase": "configurePhase" }
Running phase: configurePhase
no configure script, doing nothing
@nix { "action": "setPhase", "phase": "buildPhase" }
Running phase: buildPhase
zig build flags: -Dcpu=baseline -Doptimize=ReleaseSafe -Dpie=true
error: unable to create compilation: DarwinSdkNotFound
```

### Additional context

This seems to be the first failure on Hydra: https://hydra.nixos.org/build/247187184

These `DarwinSdkNotFound` errors also affect other packages using Zig, such as `dt`: https://hydra.nixos.org/build/248856951

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
 - system: `"x86_64-darwin"`
 - host os: `Darwin 23.3.0, macOS 10.16`
 - multi-user?: `yes`
 - sandbox: `relaxed`
 - version: `nix-env (Nix) 2.18.1`
 - nixpkgs: `/nix/store/y7k6gpjwwspbpwvw3vax3xc3y53qg0rj-source`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by Samasaur1** [comment]
ncdu builds for me with this `nix-info -m`:
```
 - system: `"aarch64-darwin"`
 - host os: `Darwin 23.3.0, macOS 14.3.1`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.18.1`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`
 ```

**Comment by Samasaur1** [comment]
You may want to ask in [#macos:nixos.org](https://matrix.to/#/#macos:nixos.org), they'd be the ones to help debug this issue

**Comment by eclairevoyant** [comment]
This seems to be a repeat of #214545 but with a different error

**Comment by al3xtjames** [comment]
While trying to bisect this, I realized that `zig` (more specifically, `zig_0_11` itself was broken - the build was failing with `OSVersionDetectionFail` during stage 3). It turns out that Zig calls `std.zig.system.darwin.macos.detect` during the build, which [parses a couple of property list files in /System/Library/CoreServices to determine the OS version][1]. This was failing as I had `sandbox = relaxed` set in nix.conf. Adding these files to `__impureHostDeps` resolved the `OSVersionDetectionFail` error, but Zig eventually hit the `DarwinSdkNotFound` error described in this PR. [Zig calls `"/usr/bin/xcode-select` and `/usr/bin/xcrun`][2] to determine the macOS SDK version and path (this is impure as it's using SDK provided by Xcode). I tried to replace them with the binaries provided by `xcbuild` (which should use the SDK provided by Nix, IIUC), but [stage 3 was unable to link due to missing symbols][3].

Of course, disabling the Nix sandbox or setting `__noChroot` works, though this doesn't address the impurity.

[1]: https://github.com/ziglang/zig/blob/cfce81f7d5f11ab93b2d5fd26df41edf967f333b/lib/std/zig/system/darwin/macos.zig#L12
[2]: https://github.com/ziglang/zig/blob/cfce81f7d5f11ab93b2d5fd26df41edf967f333b/lib/std/zig/system/darwin.zig
[3]: https://gist.github.com/al3xtjames/44e82fefd0d9aabceee4aeea10327c62

**Comment by iFreilicht** [comment]
Zig is building properly on darwin since version 0.12, which is the new pinned dependency for ncdu since https://github.com/NixOS/nixpkgs/pull/307243. [Hydra has successfully built and cached ncdu on macOS as well](https://hydra.nixos.org/eval/1806026?filter=ncdu&compare=1806018&full=), so this issue is fixed on unstable.

We won't be backporting this fix to 23.05, but as 24.05 is just around the corner, it will land in there.
